src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,698 @@
+                if (_eventProperties != null && _eventProperties.TryGetValue(key, out value))
+                    return value;
+
+                int keyIndex = MessagePropertiesIndexOfKey(key);
+                if (keyIndex != -1)
+                {
+                    return _messageProperties[keyIndex].Value;
+                }
+
+                throw new KeyNotFoundException();
+            }
+            set
+            {
+                EventProperties[key] = value;
+            }
+        }
+
+        /// <inheritDoc/>
+        public ICollection<object> Keys { get { return (ICollection<object>)KeyCollection ?? new DictionaryCollection(this, true); } }
+
+        /// <inheritDoc/>
+        public ICollection<object> Values { get { return (ICollection<object>)ValueCollection ?? new DictionaryCollection(this, false); } }
+
+        private Dictionary<object, object>.KeyCollection KeyCollection
+        {
+            get
+            {
+                if (_eventProperties != null && (_messageProperties == null || _messageProperties.Count == 0))
+                    return _eventProperties.Keys;
+                if (_eventProperties == null && (_messageProperties == null || _messageProperties.Count == 0))
+                    return EmptyCollection.Keys;
+                return null;
+            }
+        }
+
+        private Dictionary<object, object>.ValueCollection ValueCollection
+        {
+            get
+            {
+                if (_eventProperties != null && (_messageProperties == null || _messageProperties.Count == 0))
+                    return _eventProperties.Values;
+                if (_eventProperties == null && (_messageProperties == null || _messageProperties.Count == 0))
+                    return EmptyCollection.Values;
+                return null;
+            }
+        }
+
+        private static readonly Dictionary<object, object> EmptyCollection = new Dictionary<object, object>();
+
+        /// <inheritDoc/>
+        public int Count { get { return (_messageProperties != null ? _messageProperties.Count : 0) + (_eventProperties != null ? _eventProperties.Count : 0); } }
+
+        /// <inheritDoc/>
+        public bool IsReadOnly { get { return false; } }
+
+        /// <inheritDoc/>
+        ICollection IDictionary.Keys { get { return (ICollection)KeyCollection ?? new DictionaryCollection(this, true); } }
+
+        /// <inheritDoc/>
+        ICollection IDictionary.Values { get { return (ICollection)ValueCollection ?? new DictionaryCollection(this, false); } }
+
+        /// <inheritDoc/>
+        bool IDictionary.IsFixedSize { get { return false; } }
+
+        /// <inheritDoc/>
+        object ICollection.SyncRoot { get { return null; } }
+
+        /// <inheritDoc/>
+        bool ICollection.IsSynchronized { get { return false; } }
+
+        /// <inheritDoc/>
+        public void Add(object key, object value)
+        {
+            int keyIndex = MessagePropertiesIndexOfKey(key);
```

Comment:
so add is now `O(n)`?

Reply:
Only when using structured-logging, and there will only be a performance hit if using more than 100 message-parameters.